refinery/core/views.py

```diff
@@ -1015,11 +1015,82 @@ def partial_update(self, request, uuid, format=None):
+
+    def get_object(self, uuid):
+        try:
+            return ExtendedGroup.objects.get(uuid=uuid)
+        except ExtendedGroup.DoesNotExist as e:
+            logger.error(e)
+            raise Http404
+        except ExtendedGroup.MultipleObjectsReturned as e:
+            logger.error(e)
+            raise APIException("Multiple groups returned for this request.")
+
+    def get_user(self, id):
+        try:
+            return User.objects.get(id=id)
+        except User.DoesNotExist as e:
+            logger.error(e)
+            raise Http404
+        except User.MultipleObjectsReturned as e:
+            logger.error(e)
+            raise APIException("Multiple users returned for this request.")
+
+    def post(self, request, uuid):
+        # if group is a manager_group, user is promoted by adding to user set
+        group = self.get_object(uuid)
+
+        if group.is_user_a_group_manager(request.user):
+            edit_user = self.get_user(request.data.get('userId'))
+            if group.is_manager_group():
+                group.user_set.add(edit_user)
+                serializer = ExtendedGroupSerializer(group,
+                                                     context={user: edit_user})
+                return Response(serializer.data)
+            return HttpResponseBadRequest(
+                content="Manager groups are required to upgrade."
+                )
+        # not a group manager
+        return Response(uuid, status=status.HTTP_403_FORBIDDEN)
+
+    def delete(self, request, uuid, id):
+        # if group is a manager_group, user is demoted by removal
+        group = self.get_object(uuid)
+        edit_user = self.get_user(id)
+        if self.is_user_authorized_to_edit(group, request.user, edit_user):
+            return Response(uuid, status=status.HTTP_403_FORBIDDEN)
+
+        if group.id == settings.REFINERY_PUBLIC_GROUP_ID:
+            return HttpResponseBadRequest(
+                content="Users can not leave public group."
+            )
+        # Demote
         if group.is_manager_group():
-            return user in group.user_set.all()
-        else:
-            return user in group.manager_group.user_set.all()
+            if len(group.user_set.all()) > 1:
+                group.user_set.remove(edit_user)
+                return Response(uuid)
+            else:
+                return HttpResponseBadRequest(
+                    content="Last manager must delete group to leave."
+                )
+        # Leave
+        if group.is_user_a_group_manager(edit_user):
+            return HttpResponseBadRequest(
+                content="Managers can not leave group. Demote user first."
+            )
+        if len(group.user_set.all()) > 0:
+            group.user_set.remove(edit_user)
+            return Response(uuid)
+        return HttpResponseBadRequest(content="No users left in group.")
+
+    def is_user_authorized_to_edit(self, group, request_user, edit_user):
```

Comment:
Sounds like it should be a method either on the Group or the User model. This would help with reuse and simplify the function signature.

Reply:
It's only used for this delete request which is dependent on both the member and request user. I don't imagine it being used anywhere else, so it's fine here.